### DISCUSSION TITLE
Get tags for containing widget when using list

### DISCUSSION THREAD

**Discussion by welford** [TiddlyWiki/TiddlyWiki5#5852] [Q&A] [answered]
I have a widget that will hide a section of text based on, among other things, the presense of the tag "$:/tags/twsr/hideAnswers"

I do this in javascript with:

```
var ct = this.getVariable("currentTiddler");
var tid = $tw.wiki.getTiddler(ct);
var tags = tid ? tid.getFieldList("tags") : [];
if(tags.indexOf("$:/tags/twsr/hideAnswer") >= 0 || tags.indexOf("$:/tags/twsr/hideAnswers") >= 0){
	locallyActive = true;
}
...
```

But this is inadequate for the case where the widget is brought in via a list. see second tiddler here: http://welford.github.io/ruby.html

Is there a way i can make this work? get the tags for the actual tiddler being rendered and not whatever currentTiddler is here.

Thanks,
James

**Comment by welford** [comment, marked as answer]
Went with something like this, traverse parent widgets and pick up all tags along the way, would be nice if i could pick out the container widget that actually is showing though.  Is there a way to do this?

```
var p = this;
var tags = [];
while(p){
	var ct = p.getVariable("currentTiddler");
	var tid = $tw.wiki.getTiddler(ct);
	if(tid){
		tags = tags.concat(tid.getFieldList("tags"));
		tags = tags.filter(function(item, pos) {	return tags.indexOf(item) == pos; })
	}
	p = p.parentWidget;
}
```

**Reply by pmario** [reply]
IMO this is completely inefficient and slow. ... Accessing the global $tw.wiki object shouldn't be needed, except for utility functions. ... 

Where is the whole code of the widget? .. Some description, what it should do would also be nice, to see what's going on. 
I did find http://welford.github.io/ruby.html#%24%3A%2Fplugins%2Fwelford%2Ftwsr%2Ftwsr.js .. But I would need to reverse engineer your widget code, to understand it. .. 

That definitely needs more time as I want to invest.

**Reply by welford** [reply]
No worries pmario, what i am trying to do is get the tags of the currently rendering tiddler. list widget by default stomps currentTiddler. I was unaware of the  "variable" attribute in the widget, which helps work around the problem. 

The behaviour i am trying to achieve is this: if the tag "$:/tags/twsr/hideAnswers" is present, then the "ruby" element above the text should always be blacked out. e.g:

![image](https://user-images.githubusercontent.com/4590399/126986552-d2e83b92-0649-4ff4-af81-9757588218f6.png)


I think my widget solution, while maybe not ideal, probably matches behaviour that an end user of this would expect. Especially given that, while no expert, I have tinkered with TW for years now and this still stumped me :) 

I update this http://welford.github.io/ruby.html with my new widget and a new examples:

"bad" list construction, but fixed up by my widget
```
<$list filter="[tag[apple]]">
<$r u=<<currentTiddler>>><<currentTiddler>></$r>
</$list>
```

Correct use of variable:
```
<$list filter="[tag[apple]]" variable="fruit">
<$r u=<<fruit>>><<fruit>></$r>
</$list>
```

In this case `<<currentTiddler>>` still picks up the outer tiddlers tags & the widgets works as expected.

Thanks for the help.

**Reply by pmario** [reply]
You are welcome! ... 

But as I wrote. ... Once you are happy with your functionality, you may consider to improve the code. I think in the long run your maintenance time will go down. Especially if you have users, that typically don't use the stuff as intended. :)

**Comment by pmario** [comment]
Hi, .. I don't understand, why you hardcode a tag in a widget?! ... IMO that's bad practice. Every parameter, that changes the behaviour of a widget should be a real parameter.

**Reply by welford** [reply]
> I don't understand, why you hardcode a tag in a widget?!

it's just quicker and easier while developing,  The larger problem is in my original question. I'm not sure if my solution is good practice,  but it works.

**Reply by pmario** [reply]
Yea, but eg variables like this: `g_questionElm` may cause problems in the future, since they are global outside the widget instance. They need to move into 

```
var TWSRWidget = function(parseTreeNode,options) {
    this.initialise(parseTreeNode, options);
    this.g_questionElm = null;
// the rest of your global vars should be here. 
};
```

They can be used with `this.g_questionElm` in the other functions of the instance.

**Comment by pmario** [comment]
I think the original question is also a problem of missing parameters. .. If a widget needs a "tiddler" parameter it needs to have a "parameter", otherwise you will get the problems that you have now. 

The [list widget](https://tiddlywiki.com/#ListWidget) has a `variable` parameter. So if you use 

```
<$list filter="[tag[fruit]]" variable="anyName">

<$r u=<<currentTiddler>>><<currentTiddler>></$r>

</$list>
```

Your `<<currentTiddler>>` variable will be the outer variable. BUT this variable may come from an other widget or a transclusion. So your widget will fail again if used in a different context. ...

IMO you should really think about to get your parameters fixed.

**Comment by pmario** [comment]
The closest widget I can think of is the [reveal-widget](https://github.com/Jermolene/TiddlyWiki5/blob/19fd5ca5f26f798742e411453b30ad467b101a50/core/modules/widgets/reveal.js). .. So if you have a closer look at the reveal-widget code, you can see, how it handles parameters. 

The other and much more complex widget that may be interesting is the "button-widget" .. BUT it's code is a mess since it does much more things as it should ...
